Manually add lookup table addresses instead of sanitizing

apfitzge opened the pull request:
#### Problem
#30165

#### Summary of Changes
Manually load and add lookup table addresses instead of using sanitize.

Fixes #30165
<!-- OPTIONAL: Feature Gate Issue: # -->
<!-- Don't forget to add the "feature-gate" label -->

apfitzge commented:
Was able to test this with a script that spammed creating and using an ALT. This makes it so that some ALTs used in txs do not exist at our beginning state but are used in a transaction. That transaction cannot be sanitized with the state the beginning slot bank has.

To create the snapshot: `./cargo run --release --bin solana-ledger-tool -- --ledger ./test-ledger create-snapshot --minimized --ending-slot 670 607`

Master:

```text
[2023-09-18T19:01:58.431801175Z INFO  solana_core::accounts_hash_verifier] AccountsHashVerifier has stopped
[2023-09-18T19:01:58.436755228Z INFO  solana_metrics::metrics] datapoint: calculate_accounts_hash_from_index accounts_scan=366i hash=278i hash_total=402i collect=2470i
thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: AddressLookupTableNotFound', thread 'ledger/src/blockstore.rs<unnamed>:' panicked at '2943called `Result::unwrap()` on an `Err` value: AddressLookupTableNotFound:', 82ledger/src/blockstore.rs
:note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
2943:82
thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: AddressLookupTableNotFound', ledger/src/blockstore.rs:2943:82
thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: AddressLookupTableNotFound', ledger/src/blockstore.rs:2943:82
thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: AddressLookupTableNotFound', ledger/src/blockstore.rsthread ':<unnamed>2943' panicked at ':called `Result::unwrap()` on an `Err` value: AddressLookupTableNotFound82',
ledger/src/blockstore.rs:2943:82
thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: AddressLookupTableNotFound', ledger/src/blockstore.rs:2943:82
thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: AddressLookupTableNotFound', ledger/src/blockstore.rs:2943:82
thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: AddressLookupTableNotFound', ledger/src/blockstore.rs:2943:82
thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: AddressLookupTableNotFound', ledger/src/blockstore.rs:2943:82
```

Branch:

```text
[2023-09-18T19:11:44.732909031Z INFO  solana_ledger_tool] ledger tool took 1.4s
```

ryoqun commented on ledger/src/blockstore.rs:
```diff
@@ -2922,37 +2924,44 @@ impl Blockstore {
+            };
+
+            for address in lookup_table.addresses.iter() {
+                result.insert(*address);
```
i wish `ExtendLookupTable` accepted addresses in the form of account references, not instruction data with serialized `Vec<Pubkey>`....

if so, things would be simpler...

apfitzge replied:
Oh didn't realize it was through ix data.
I can scan for extend ixs and manually add those as well

apfitzge replied:
> i wish ExtendLookupTable accepted addresses in the form of account references, not instruction data with serialized Vec<Pubkey>....

Agree this would make things much simpler (for us). But I can understand why it was not doen that way - since there's no need to load those, possibly large, accounts to add them to an ALT.

apfitzge replied:
> how about plumbing a flag --non-strict-alt-scanning and panic! as before by default with mention of the flag?

I think these are common enough that I'd like to consider some alternatives first, because otherwise users will use default and hit this often, or have to run twice.

I'm more in favor of doing our best, and logging a warning if we detect a possible issue...which I think is possible?

What do you think about this:
1. We manually scan for "simple" extension instructions
  - any txs w/ native-only programs/ixs
  - we can directly rip pubkeys from the extension ix data  
3. We try detecting potential CPI:
  - even if they CPI the ALT extension, the tx must have the ALT program as an input
  - if we see some tx which uses non-native programs & has the ALT program this means it's possible it will CPI to extend.
  - if we detect this case, we log some warning (at the end) with a description
  
I think this makes the usage a lot easier and robust than it currently is. And it still allows us to print a warning that it may not work, rather than just silently not working.
I'm not sure how common CPI extensions are, but I think we certainly can do better for "simple" ALT create & extends.

ryoqun replied:
> @ryoqun what do you think about this:
> ...

sounds nice. a little concern is that people won't notice those warnings as `ledger-tool create-snapshot` is already quite verbose. maybe we can arrange to display the warning at the very last of command termination for maximum discoverability as you said.

apfitzge commented on sdk/program/src/lib.rs:
```diff
@@ -585,6 +585,9 @@ pub mod sdk_ids {
                 vote::program::id(),
                 feature::id(),
                 bpf_loader_deprecated::id(),
+                address_lookup_table::program::id(),
+                loader_v4::id(),
+                stake::program::id(),
```
found some missing native programs that should have always been in this list 😦

apfitzge commented:
Rebased on master to fix the cargo audit for `quinn-proto 0.10.4`

ryoqun commented on ledger/src/transaction_address_lookup_table_scanner.rs:
```diff
@@ -0,0 +1,64 @@
+        if address_lookup_table::program::check_id(program_id) {
+            if let ProgramInstruction::ExtendLookupTable { new_addresses } =
+                deserialize::<ProgramInstruction>(&instruction.data)
+                    .expect("invalid address lookup table instruction")
```
maybe, we should `warn!` and ignore the `instruction` when `deserialize()` fails, instead of `.expect()` (i.e. panic). anyone can freely submit txes with malformed ix data for the address_lookup_table program.

apfitzge replied:
Yes, you're right. These txs would fail, but they can still be in our ledger/history, which this current code fails to handle.

apfitzge replied:
I think logging warning on this would be unnecessary. If we're failing to deserialize the instruction, then that transaction won't succeed and shouldn't matter anyway?

ryoqun commented on ledger-tool/src/main.rs:
```diff
@@ -3245,6 +3249,10 @@ fn main() {
                                 }
+
+                                if minimize_snapshot_possibly_incomplete {
+                                    warn!("Minimized snapshot may be incomplete due to missing accounts from CPI'd address lookup table extensions. This may lead to replay failures");
```
`This may lead to replay failures` => `This may lead to mismatched bank hashes while replaying`. Missing actually accessed accounts in minimized snapshot won't cause hard errors as indicated by `replay failures`. it's more subtle and scary bank hash mismatches out of sudden. that's because those missed accounts are just filled `Account::default()` and on-chain programs just start to behave differently. (most likely just erring..)

ryoqun commented on ledger/src/blockstore.rs:
```diff
@@ -2915,46 +2918,70 @@ impl Blockstore {
+
+                            let tx = SanitizedVersionedTransaction::try_from(tx)
+                                .expect("transaction failed to sanitize");
+                            let alt_scan_result =
+                                transaction_address_lookup_table_scanner::scan_transaction(&tx);
+                                match alt_scan_result {
+                                    transaction_address_lookup_table_scanner::ScanResult::NotFound => {},
+                                    transaction_address_lookup_table_scanner::ScanResult::NativeUsed(keys) => {
+                                        keys.into_iter().for_each(|key| { result.insert(key); });
+                                    }
+                                    transaction_address_lookup_table_scanner::ScanResult::NonNativeUsed(keys) => {
```
nit: how about `use transaction_address_lookup_table_scanner::{...}`-ing?

apfitzge replied:
Yeah didn't do that originally because the `ScanResult` name was terrible and too generic. Now that that is going to change I'll just `use` here

ryoqun commented on ledger/src/blockstore.rs:
```diff
@@ -2915,46 +2918,70 @@ impl Blockstore {
+            }
+        });
+
+        (result, possible_cpi_alt_extend.load(Ordering::Relaxed))
```
nit: `into_inner()` is shorter and semantically more straightforward: https://doc.rust-lang.org/std/sync/atomic/struct.AtomicBool.html#method.into_inner

ryoqun commented on ledger/src/blockstore.rs:
```diff
@@ -2915,46 +2918,70 @@ impl Blockstore {
+                return;
+            };
+
+            for address in lookup_table.addresses.iter() {
```
nit: mixed use of `for` loops and `.for_each()` in this pr to do `result.insert(...)`. maybe stick to one of the two?

ryoqun commented on ledger/src/blockstore.rs:
```diff
@@ -2915,46 +2918,70 @@ impl Blockstore {
-        result
+        // For each unique lookup table add all accounts to the minimized set.
+        lookup_tables.into_par_iter().for_each(|lookup_table_key| {
+            let Some(lookup_table_account) = bank.get_account(&lookup_table_key) else {
+                return;
+            };
+
+            let Ok(lookup_table) = AddressLookupTable::deserialize(lookup_table_account.data())
+            else {
+                return;
+            };
+
+            for address in lookup_table.addresses.iter() {
+                result.insert(*address);
+            }
```
nit: i'd personally prefer `.and_then()` chaining.

apfitzge replied:
Chained `and_then` doesn't work well because the `AddressLookupTable` has references into the data. I can be more concise though.

ryoqun commented on ledger/src/transaction_address_lookup_table_scanner.rs:
```diff
@@ -0,0 +1,64 @@
+
+pub fn scan_transaction(transaction: &SanitizedVersionedTransaction) -> ScanResult {
+    // If the program is not present in the account keys, it was not used
+    if !transaction
+        .get_message()
+        .message
+        .static_account_keys()
+        .iter()
+        .any(address_lookup_table::program::check_id)
+    {
+        return ScanResult::NotFound;
+    }
```
sadly, this isn't precise.. the alt could be loaded via another alt. this is okay as long as non-cpi ix isn't touching the alt-loaded alt program. cpi ix is allowed to execute alt-loaded on-chain programs.

ryoqun replied:
how about SanitizedTransaction construction as was before then only fall back to `static_account_keys()`?

this will reduce otherwise unavoidable false positive warnings greatly...

```rust
let account_keys = bank
    .fully_verify_transaction(transaction)
    .map(|tx| tx.message.account_keys()).unwrap_or_else(|| {
        possibly_incomplete = true;
        transaction.get_message().message.static_account_key()
    });
if !possibly_incompete &&
    account_keys.iter().any(address_lookup_table::program::check_id) {
    return ScanResult::NotFound;
}
...
```

apfitzge replied:
I did something similar, outside the scanning.
i.e. we only scan if we cannot verify.

However...it requires cloning the `tx` because `fully_verify_transaction` consumes.

ryoqun reviewed (commented):
thanks for working on this...

I'm moderately sure you're starting to realize why I didn't do this originally. xD

ryoqun commented on ledger/src/transaction_address_lookup_table_scanner.rs:
```diff
@@ -0,0 +1,64 @@
+    static ref SDK_IDS_SET: HashSet<Pubkey> = SDK_IDS.iter().cloned().collect();
+}
+
+pub enum ScanResult {
```
naming: `Result` is too generic to add any valuable context. also possibly incomplete vs (non)native naming abstraction is too much for the amount of relevant code, just adding cognitive load... also, the two variant shared the same type shape..

how about this?


```
pub struct ScannedExtension(s) {
    possibly_incomplete: bool,
    extended_keys: Vec<Pubkey>,
}
```

ryoqun commented on ledger/src/blockstore.rs:
```diff
@@ -2915,46 +2918,68 @@ impl Blockstore {
+        let lookup_tables = DashSet::new();
+        let possible_cpi_alt_extend = AtomicBool::new(false);
+
+        fn add_to_set<'a>(set: &DashSet<Pubkey>, iter: impl IntoIterator<Item = &'a Pubkey>) {
```
nice refactoring.

nit: you can create closure not to pass around `result` maybe? like:

```rust
let add_to_set = |iter: ...| { ... result.insert(...) ... } 
```

ryoqun replied:
oh, this is used for `result` and `lookup_tables`.. never mind...

ryoqun commented on ledger/src/transaction_address_lookup_table_scanner.rs:
```diff
@@ -0,0 +1,56 @@
+pub fn scan_transaction(
+    transaction: &SanitizedVersionedTransaction,
+) -> ScannedLookupTableExtensions {
+    // If the ALT program is not present in the account keys, it was not used
+    if !transaction
+        .get_message()
+        .message
+        .static_account_keys()
+        .iter()
+        .any(address_lookup_table::program::check_id)
+    {
+        return ScannedLookupTableExtensions::default();
+    }
```
i think we still need to return `possibly_incomplete: true` for the possibility of

ryoqun commented on ledger/src/transaction_address_lookup_table_scanner.rs:
```diff
@@ -0,0 +1,56 @@
+        .iter()
+        .any(address_lookup_table::program::check_id)
+    {
+        return ScannedLookupTableExtensions::default();
```
nit: this is really edge of edge case.

but, i think we still need to return from here with `possibly_incomplete: true` for the possibility of alt extension via cpi, even considering that this scanning now only occurs when `fully_verify_transaction()` failed.

that's because this is possible unless all outermost (non-cpi) iexes are of native programs.

how about removing this `if` itself and remove `#[derive(Default)]` from `ScannedLookupTableExtensions`?

ryoqun reviewed (dismissed):
lgtm with nit

thanks for improving on my past lazy work.

although i wrote somewhat long nit code review comment, i don't want to block this pr anymore for my super picky nit. ;)

apfitzge commented:
> although i wrote somewhat long nit code review comment, i don't want to block this pr anymore for my super picky nit. ;)

I removed the `if` in the function, and `Default` impl. Only change since your last review...however, I rebased since there were a few cargo vulnerabilities which were fixed on master and flagging the CI.

ryoqun reviewed (approved):
lgtm

much much better. :)

thanks very much for addressing multi round of code reviews..
